src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,382 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+  size_t left = 0;
+  size_t right = length;
+  size_t bestFit = 0;
+
+  while (left <= right) {
+    size_t mid = left + (right - left) / 2;
+    if (mid == 0) break;
+
+    // Don't split surrogate pairs - adjust backwards if mid lands after a high surrogate
+    size_t adjustedMid = mid;
+    if (adjustedMid > 0 && adjustedMid < length) {
+      char16_t prev = data[adjustedMid - 1];
+      if (prev >= 0xD800 && prev < 0xDC00) {
+        adjustedMid--;
+      }
+    }
+
+    if (adjustedMid == 0) {
+      right = 0;
+      break;
+    }
+
+    size_t midUtf8Length = simdutf::utf8_length_from_utf16(data, adjustedMid);
+    if (midUtf8Length <= bufferSize) {
+      bestFit = adjustedMid;
+      left = adjustedMid + 1;
+    } else {
+      right = adjustedMid - 1;
+    }
+  }
+
+  return bestFit;
+}
+
+// Binary search to find how many UTF-16 code units with invalid surrogates fit when converted to UTF-8.
+// Ensures surrogate pairs are never split, and unpaired surrogates are replaced with U+FFFD.
+size_t findBestFitInvalidUtf16(const char16_t* data, size_t length, size_t bufferSize) {
+  size_t left = 0;
+  size_t right = length;
+  size_t bestFit = 0;
+
+  while (left <= right) {
+    size_t mid = left + (right - left) / 2;
+    if (mid == 0) break;
+
+    // Don't split surrogate pairs - adjust backwards if mid lands after a high surrogate
+    size_t adjustedMid = mid;
+    if (adjustedMid > 0 && adjustedMid < length) {
+      char16_t prev = data[adjustedMid - 1];
+      if (prev >= 0xD800 && prev < 0xDC00) {
+        adjustedMid--;
+      }
+    }
+
+    if (adjustedMid == 0) {
+      right = 0;
+      break;
+    }
+
+    size_t midUtf8Length = utf8LengthFromInvalidUtf16(kj::arrayPtr(data, adjustedMid));
+    if (midUtf8Length <= bufferSize) {
+      bestFit = adjustedMid;
+      left = adjustedMid + 1;
+    } else {
+      right = adjustedMid - 1;
+    }
+  }
+
+  return bestFit;
 }
 
+}  // namespace
+
 TextEncoder::EncodeIntoResult TextEncoder::encodeInto(
     jsg::Lock& js, jsg::JsString input, jsg::JsUint8Array buffer) {
-  auto result = input.writeInto(
-      js, buffer.asArrayPtr<char>(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
+  auto outputBuf = buffer.asArrayPtr<char>();
+  size_t bufferSize = outputBuf.size();
+
+  v8::String::ValueView view(js.v8Isolate, input);
+  uint32_t length = view.length();
+
+  if (view.is_one_byte()) {
+    // Latin-1 path: characters 0x00-0x7F encode as 1 UTF-8 byte, 0x80-0xFF as 2 bytes
+    auto data = reinterpret_cast<const char*>(view.data8());
+
+    // Fast path: avoid length calculation when we can prove the string fits.
+    // Check worst-case (2x), ASCII (1:1), or calculate exact length as fallback.
+    size_t read = length;
+    if (!(length * 2 <= bufferSize ||
+            (length <= bufferSize && simdutf::validate_ascii(data, length)) ||
+            simdutf::utf8_length_from_latin1(data, length) <= bufferSize)) {
```

Comment:
nit: I think our allowed lengths for strings are small enough that it shouldn't be a problem but there's at least a theoretical risk of overflow in this check... similarly in a few other places (line 819 below for instance).

Reply:
Resolving due to Erik's comment about this being safe.